Source/WebCore/rendering/GridMasonryLayout.cpp

```diff
@@ -185,10 +206,19 @@ GridArea GridMasonryLayout::nextMasonryPositionForItem(const RenderBox* item)
             smallestMaxPosLine = startingLine;
         } 
     }
+    return GridSpan::translatedDefiniteGridSpan(smallestMaxPosLine, smallestMaxPosLine + itemSpanLength);
+}
 
-    GridSpan masonryAxisSpan = GridSpan::translatedDefiniteGridSpan(0, 1);
-    GridSpan gridAxisSpan = GridSpan::translatedDefiniteGridSpan(smallestMaxPosLine, smallestMaxPosLine + itemSpanLength);
-    return m_masonryAxisDirection == ForRows ? GridArea { masonryAxisSpan, gridAxisSpan } : GridArea { gridAxisSpan, masonryAxisSpan };
+GridSpan GridMasonryLayout::gridAxisPositionUsingNextAutoFlow(const RenderBox* item) const
+{
+    return GridSpan::translatedDefiniteGridSpan(m_autoFlowNextCursor, m_autoFlowNextCursor + GridPositionsResolver::spanSizeForAutoPlacedItem(*item, gridAxisDirection()));
```

Comment:
What happens in the case where the item has a definite span that is larger than the explicit grid, and m_autoFlowNextCursor is not 0.

For example, say we have (in the grid axis), 3 explicit columns, and an item that has requested 'span 5'.

The initial 'ignoring masonry' placement step should have run Step 3 of the [Grid Item placement algorithm](https://drafts.csswg.org/css-grid-2/#auto-placement-algo), which will determine that we need 2 extra implicit columns for that item to fit. I'd expect m_gridAxisTracksCount to be 5 as a result of that.

If we're doing next placement, and m_autoFlowNextCursor is not-0, then this code will result in the item extending past the end of the implicit grid.
Inserting in the grid modifies the implicit grid, but it won't change m_gridAxisTracksCount, and they'll be out of sync.

I can't tell what the spec wants here. For non-masonry placement, we'd reset back to 0 if there wasn't enough room on the line to fit the item, rather than extending the implicit grid a second time.

Reply:
From section 2.3: " plus the items that span all grid axis tracks (which obviously will be placed at the first implicit line since there is no other way to place them). "

Reply:
Yeah I think it makes sense for the cursor to reset back to the beginning of the grid if the item would span out of the grid. I added a couple of test cases to cover this scenario.